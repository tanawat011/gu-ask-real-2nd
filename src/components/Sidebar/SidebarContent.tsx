import tw, { styled } from 'twin.macro'

import { FlexCol } from 'components/Common'

export type MenuItem = {
  label: string
  to: string
}

export type SidebarContentProps = {
  menuList: MenuItem[]
  itemSelected: string
  handleClickItem: (key: string, item: MenuItem) => void
}

type TwSidebarContentProps = Partial<SidebarContentProps> & {
  isActive?: boolean
}

const Container = tw(FlexCol)`h-full w-full bg-space-blue pt-[26px]`
const Item = styled.div(({ isActive }: TwSidebarContentProps) => [
  tw`relative flex cursor-pointer flex-row items-center h-[40px] mb-[12px]`,
  isActive && tw`bg-iron-gray`,
])
const ItemActive = tw.div`absolute z-0 h-full rounded-r bg-toy-red w-[6px]`
const ItemLabel = styled.div(({ isActive }: TwSidebarContentProps) => [
  tw`font-thaisans-nue z-0 font-semibold text-iron-gray text-[18px] ml-[22px] pb-[6px]`,
  isActive && tw`text-white`,
])

export const SidebarContent: React.FC<SidebarContentProps> = ({
  menuList,
  itemSelected,
  handleClickItem,
}) => {
  return (
    <Container>
      {menuList.map((item, index) => {
        const keyName = item.to.split('/')[1]
        const isActive = keyName === itemSelected

        return (
          <Item
            key={`sidebar-menu-item-${index}`}
            onClick={() => handleClickItem(keyName, item)}
            isActive={isActive}
          >
            {isActive && <ItemActive />}

            <ItemLabel isActive={isActive}>{item.label}</ItemLabel>
          </Item>
        )
      })}
    </Container>
  )
}
