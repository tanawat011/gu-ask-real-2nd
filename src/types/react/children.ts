export type Children = { children: JSX.Element }
