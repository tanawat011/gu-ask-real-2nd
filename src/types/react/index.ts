export * from './children'
export * from './reactRouterDom'
