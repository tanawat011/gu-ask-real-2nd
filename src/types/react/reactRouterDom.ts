import type { Scroll } from 'types/common'

export type OutletContextProps = {
  scroll: Scroll
}
