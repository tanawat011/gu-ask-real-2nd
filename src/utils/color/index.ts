export * from './hex2Rgb'
export * from './rgb2Hex'
