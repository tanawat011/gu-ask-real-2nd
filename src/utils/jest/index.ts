export * from './renderWithProviders'
export * from './twColor'
