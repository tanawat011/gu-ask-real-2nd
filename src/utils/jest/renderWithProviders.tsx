import type { RenderResult } from '@testing-library/react'

import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { render } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import { RecoilRoot } from 'recoil'

const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  const queryClient = new QueryClient()

  return (
    <QueryClientProvider client={queryClient}>
      <RecoilRoot>
        <BrowserRouter>{children}</BrowserRouter>
      </RecoilRoot>
    </QueryClientProvider>
  )
}

interface RenderWithProvidersOptions {
  router?: {
    route: string
  }
  recoil?: object
}

export const renderWithProviders = (
  ui: JSX.Element,
  option?: RenderWithProvidersOptions,
): RenderResult => {
  const routerOption = { ...(option?.router || { route: '/' }) }
  window.history.pushState({}, 'Test page', routerOption.route)

  return {
    ...render(ui, { wrapper: AllTheProviders, ...routerOption }),
  }
}
