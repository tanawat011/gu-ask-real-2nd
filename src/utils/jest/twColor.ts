import { hex2Rgb } from 'utils/color'

export const twColor = (hexColor: string, type?: 'bg' | 'border' | 'text') => {
  if (hexColor === 'transparent') {
    return 'transparent'
  }

  if (hexColor.length > 7) {
    return 'undefined'
  }

  const rgb = hex2Rgb(hexColor, { isArray: true }) as string[]

  if (!type) {
    type = 'bg'
  }

  return `rgb(${rgb.join(' ')} / var(--tw-${type}-opacity))`
}
