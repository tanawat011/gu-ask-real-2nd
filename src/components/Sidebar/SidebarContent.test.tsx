import type { MenuItem } from './Sidebar'

import { fireEvent, screen, waitFor } from '@testing-library/react'

import { renderWithProviders } from 'utils/jest'

import { SidebarContent } from './SidebarContent'

jest.mock('./SidebarContent', () => jest.requireActual('./SidebarContent'))
jest.mock('./SidebarItem', () => ({
  SidebarItem: ({ onClick }: { onClick: () => void }) => (
    <div onClick={onClick}>sidebar item</div>
  ),
}))
jest.mock('./SidebarItemChild', () => ({
  SidebarItemChild: ({ onClick }: { onClick: () => void }) => (
    <div onClick={onClick}>sidebar item child</div>
  ),
}))

let menuList: MenuItem[] = [
  {
    title: 'APPS',
    children: [
      {
        to: '/todo',
        label: 'Todo',
        icon: <div>xxx</div>,
      },
      {
        label: 'Article',
        icon: <div>xxx</div>,
        expanded: true,
        children: [
          {
            to: '/sub-article',
            label: 'Sub Article',
          },
        ],
      },
    ],
  },
]

let menuListNoRoute: MenuItem[] = [
  {
    title: 'APPS',
    children: [
      {
        to: '',
        label: 'Todo',
        icon: <div>xxx</div>,
      },
      {
        label: 'Article',
        icon: <div>xxx</div>,
        children: [
          {
            to: '',
            label: 'Sub Article',
          },
        ],
      },
    ],
  },
]

describe('<SidebarContent />', () => {
  test('should render without crashing and match snapshot', async () => {
    expect.hasAssertions()

    const setMenuList = (newList: MenuItem[]) => {
      menuList = newList
    }

    const { asFragment } = renderWithProviders(
      <SidebarContent
        fullSidebar
        menuList={menuList}
        setMenuList={setMenuList}
      />,
    )

    const item = screen.getAllByText(/sidebar item/i)
    fireEvent.click(item[0])
    fireEvent.click(item[1])

    const itemChild = screen.getByText(/sidebar item child/i)
    fireEvent.click(itemChild)

    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot()
    })
  })

  test('should render without crashing and mini sidebar', async () => {
    expect.hasAssertions()

    const setMenuList = (newList: MenuItem[]) => {
      menuList = newList
    }

    const { asFragment } = renderWithProviders(
      <SidebarContent
        fullSidebar={false}
        menuList={menuList}
        setMenuList={setMenuList}
      />,
    )

    const item2 = screen.getAllByText(/sidebar item/i)[1]
    fireEvent.click(item2)

    const itemChild = screen.getByText(/sidebar item child/i)
    fireEvent.click(itemChild)

    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot()
    })
  })

  test('should render without crashing and full sidebar but no route', async () => {
    expect.hasAssertions()

    const setMenuList = (newList: MenuItem[]) => {
      menuListNoRoute = newList
    }

    const { asFragment } = renderWithProviders(
      <SidebarContent
        fullSidebar
        menuList={menuListNoRoute}
        setMenuList={setMenuList}
      />,
    )

    const item = screen.getAllByText(/sidebar item/i)
    fireEvent.click(item[0])
    fireEvent.click(item[1])

    const itemChild = screen.getByText(/sidebar item child/i)
    fireEvent.click(itemChild)

    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot()
    })
  })

  test('should render without crashing and mini sidebar but no route', async () => {
    expect.hasAssertions()

    const setMenuList = (newList: MenuItem[]) => {
      menuListNoRoute = newList
    }

    const { asFragment } = renderWithProviders(
      <SidebarContent
        fullSidebar={false}
        menuList={menuListNoRoute}
        setMenuList={setMenuList}
      />,
    )

    const item2 = screen.getAllByText(/sidebar item/i)[1]
    fireEvent.click(item2)

    const itemChild = screen.getByText(/sidebar item child/i)
    fireEvent.click(itemChild)

    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot()
    })
  })
})
