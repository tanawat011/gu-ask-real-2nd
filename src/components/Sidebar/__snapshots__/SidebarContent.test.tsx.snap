// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SidebarContent /> should render without crashing and full sidebar but no route 1`] = `
<DocumentFragment>
  .c0 {
  height: calc(100% - 64px - 64px - 1px);
}

.c1 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  padding-left: 1rem;
  padding-right: 1rem;
  padding-bottom: 1rem;
  font-size: 0.875rem;
  line-height: 1.25rem;
  font-weight: 600;
  overflow-y: overlay;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  margin-top: 1rem;
  margin-bottom: 0.5rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
}

.c4 {
  overflow: hidden;
  opacity: 1;
  -webkit-transition-property: all;
  transition-property: all;
  -webkit-transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  height: 0px;
  opacity: 0;
}

<div
    class="c0 c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        APPS
      </div>
      <div
        class="popover inline-block relative"
      >
        <div>
          sidebar item
        </div>
      </div>
      <div
        class="popover inline-block relative"
      >
        <div>
          sidebar item
        </div>
        <div
          class="c4"
        >
          <div>
            sidebar item child
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<SidebarContent /> should render without crashing and match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  height: calc(100% - 64px - 64px - 1px);
}

.c1 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  padding-left: 1rem;
  padding-right: 1rem;
  padding-bottom: 1rem;
  font-size: 0.875rem;
  line-height: 1.25rem;
  font-weight: 600;
  overflow-y: overlay;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  margin-top: 1rem;
  margin-bottom: 0.5rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
}

.c4 {
  overflow: hidden;
  opacity: 1;
  -webkit-transition-property: all;
  transition-property: all;
  -webkit-transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  height: 40px;
  margin-bottom: 0.5rem;
}

<div
    class="c0 c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        APPS
      </div>
      <div
        class="popover inline-block relative"
      >
        <div>
          sidebar item
        </div>
      </div>
      <div
        class="popover inline-block relative"
      >
        <div>
          sidebar item
        </div>
        <div
          class="c4"
        >
          <div>
            sidebar item child
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<SidebarContent /> should render without crashing and mini sidebar 1`] = `
<DocumentFragment>
  .c3 {
  width: 0;
  height: 0;
  border: 7px solid transparent;
  border-top: 7px solid;
  border-right: 7px solid;
}

.c4 {
  --tw-rotate: -135deg;
  -webkit-transform: translate(var(--tw-translate-x),var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));
  -ms-transform: translate(var(--tw-translate-x),var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));
  transform: translate(var(--tw-translate-x),var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));
  --tw-text-opacity: 1;
  color: rgb(0 0 0 / var(--tw-text-opacity));
}

.c0 {
  height: calc(100% - 64px - 64px - 1px);
}

.c1 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  padding-left: 1rem;
  padding-right: 1rem;
  padding-bottom: 1rem;
  font-size: 0.875rem;
  line-height: 1.25rem;
  font-weight: 600;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
    class="c0 c1"
  >
    <div
      class="c2"
    >
      <div
        class="popover inline-block relative"
      >
        <div>
          sidebar item
        </div>
        <div
          class="popover-content hidden absolute left-full top-1/2 -translate-y-1/2 z-10"
        >
          <div
            class="relative"
          >
            <i
              class="c3 c4 absolute left-1 top-1/2 -translate-y-1/2"
            />
            <div
              class="bg-black ml-2 text-white p-2 rounded-lg shadow-lg mb-2"
            >
              Todo
            </div>
          </div>
        </div>
      </div>
      <div
        class="popover inline-block relative"
      >
        <div>
          sidebar item
        </div>
        <div
          class="popover-content hidden absolute left-full top-1/2 -translate-y-1/2 z-10"
        >
          <div
            class="relative"
          >
            <i
              class="c3 c4 absolute text-charcoal left-1 top-1/2 -translate-y-1/2"
            />
            <div
              class="bg-charcoal ml-2 text-white rounded-lg shadow-lg mb-2"
            >
              <div>
                sidebar item child
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<SidebarContent /> should render without crashing and mini sidebar but no route 1`] = `
<DocumentFragment>
  .c3 {
  width: 0;
  height: 0;
  border: 7px solid transparent;
  border-top: 7px solid;
  border-right: 7px solid;
}

.c4 {
  --tw-rotate: -135deg;
  -webkit-transform: translate(var(--tw-translate-x),var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));
  -ms-transform: translate(var(--tw-translate-x),var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));
  transform: translate(var(--tw-translate-x),var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));
  --tw-text-opacity: 1;
  color: rgb(0 0 0 / var(--tw-text-opacity));
}

.c0 {
  height: calc(100% - 64px - 64px - 1px);
}

.c1 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  padding-left: 1rem;
  padding-right: 1rem;
  padding-bottom: 1rem;
  font-size: 0.875rem;
  line-height: 1.25rem;
  font-weight: 600;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
    class="c0 c1"
  >
    <div
      class="c2"
    >
      <div
        class="popover inline-block relative"
      >
        <div>
          sidebar item
        </div>
        <div
          class="popover-content hidden absolute left-full top-1/2 -translate-y-1/2 z-10"
        >
          <div
            class="relative"
          >
            <i
              class="c3 c4 absolute left-1 top-1/2 -translate-y-1/2"
            />
            <div
              class="bg-black ml-2 text-white p-2 rounded-lg shadow-lg mb-2"
            >
              Todo
            </div>
          </div>
        </div>
      </div>
      <div
        class="popover inline-block relative"
      >
        <div>
          sidebar item
        </div>
        <div
          class="popover-content hidden absolute left-full top-1/2 -translate-y-1/2 z-10"
        >
          <div
            class="relative"
          >
            <i
              class="c3 c4 absolute text-charcoal left-1 top-1/2 -translate-y-1/2"
            />
            <div
              class="bg-charcoal ml-2 text-white rounded-lg shadow-lg mb-2"
            >
              <div>
                sidebar item child
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
